x-pack/plugins/apm/server/routes/source_maps/schedule_source_map_migration.ts

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { ElasticsearchClient } from '@kbn/core-elasticsearch-server';
+import { FleetStartContract } from '@kbn/fleet-plugin/server';
+import { FleetArtifactsClient } from '@kbn/fleet-plugin/server/services';
+import { TaskManagerSetupContract } from '@kbn/task-manager-plugin/server';
+import { CoreStart, Logger } from '@kbn/core/server';
+import { getApmArtifactClient } from '../fleet/source_maps';
+import { bulkCreateApmSourceMaps } from './bulk_create_apm_source_maps';
+import { APM_SOURCE_MAP_INDEX } from '../settings/apm_indices/get_apm_indices';
+import { ApmSourceMap } from './create_apm_source_map_index_template';
+import { APMPluginStartDependencies } from '../../types';
+import { createApmSourceMapIndexTemplate } from './create_apm_source_map_index_template';
+
+const PER_PAGE = 10;
+const TASK_ID = 'apm-source-map-migration-task-id';
+const TASK_TYPE = 'apm-source-map-migration-task';
+
+export async function scheduleSourceMapMigration({
+  coreStartPromise,
+  pluginStartPromise,
+  fleetStartPromise,
+  taskManager,
+  logger,
+}: {
+  coreStartPromise: Promise<CoreStart>;
+  pluginStartPromise: Promise<APMPluginStartDependencies>;
+  fleetStartPromise?: Promise<FleetStartContract>;
+  taskManager?: TaskManagerSetupContract;
+  logger: Logger;
+}) {
+  if (!taskManager) {
+    return;
+  }
+
+  logger.debug(`Register task "${TASK_TYPE}"`);
+
+  taskManager.registerTaskDefinitions({
+    [TASK_TYPE]: {
+      title: 'Migrate fleet source map artifacts',
+      description:
+        'Migrates fleet source map artifacts to `.apm-source-map` index',
+      timeout: '24h',
+      maxAttempts: 5,
+      maxConcurrency: 1,
+      createTaskRunner(context) {
+        // const { state } = context.taskInstance;
+        // logger.debug({ state });
+        return {
+          async run() {
+            logger.debug(`Run task: "${TASK_TYPE}"`);
+            const coreStart = await coreStartPromise;
+            const internalESClient =
+              coreStart.elasticsearch.client.asInternalUser;
+
+            // ensure that the index template has been created before running migration
+            await createApmSourceMapIndexTemplate({
+              client: internalESClient,
+              logger,
+            });
+
+            const fleet = await fleetStartPromise;
+            if (fleet) {
+              await runFleetSourcemapArtifactsMigration({
+                fleet,
+                internalESClient,
+                logger,
+              });
+            }
+          },
+
+          async cancel() {
```

Comment:
Are there any steps we can take to actually cancel the task if the timeout is reached? Since it's performing an ES query, we could send an abort signal to the ES client perhaps?

Reply:
Is there prior art for this? If I can simply call a abort command I'm all for it. But I'd hate to complicate this by introducing a pub/sub eventing mechanism that will likely introduce more bugs than it solves.